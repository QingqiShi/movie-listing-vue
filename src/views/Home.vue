<template>
  <div class="home">
    <action-bar />
    <movie-list />
    <error-modal v-if="error">{{ error }}</error-modal>
  </div>
</template>

<script>
// @ is an alias to /src
import MovieList from "@/components/MovieList.vue";
import ActionBar from "@/components/ActionBar.vue";
import ErrorModal from "@/components/ErrorModal.vue";

export default {
  name: "Home",
  components: {
    ActionBar,
    MovieList,
    ErrorModal
  },
  props: {
    query: {
      type: Object,
      default: () => ({})
    }
  },
  computed: {
    error() {
      return this.$store.state.movies.error;
    }
  },
  mounted() {
    // Restore url query
    const { sortOrder, sortProperty, ratingFilter, genreFilter } = this.query;

    if (sortOrder !== undefined || sortProperty !== undefined) {
      this.$store.commit("setSort", {
        property:
          sortProperty !== undefined
            ? sortProperty
            : this.$store.state.movies.sortProperty,
        order:
          sortOrder !== undefined
            ? sortOrder
            : this.$store.state.movies.sortOrder
      });
    }

    if (ratingFilter !== undefined) {
      this.$store.commit("setFilter", {
        type: "rating",
        value: ratingFilter
      });
    }

    if (genreFilter) {
      genreFilter.forEach(genreId => {
        this.$store.commit("setFilter", {
          type: "genre",
          value: genreId
        });
      });
    }
  }
};
</script>

<style lang="scss">
.home {
  padding-top: 6rem;

  @include breakpoint(tablet) {
    padding-top: 3rem;
  }
}
</style>
