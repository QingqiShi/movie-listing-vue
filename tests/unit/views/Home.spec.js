import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Home from "@/views/Home.vue";

const localVue = createLocalVue();
localVue.use(Vuex);

const fakeStore = {
  modules: {
    movies: {
      state: { error: "", sortOrder: false, sortProperty: "popularity" }
    }
  },
  mutations: {
    setSort: jest.fn(),
    setFilter: jest.fn()
  }
};

describe("Home.vue", () => {
  let store;
  let wrapper;

  beforeAll(() => {
    store = new Vuex.Store(fakeStore);
    wrapper = shallowMount(Home, { store, localVue });
  });

  it("renders no error", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("renders error", () => {
    store.state.movies.error = "test error";
    expect(wrapper).toMatchSnapshot();
  });

  it("doesn't commit", () => {
    expect(fakeStore.mutations.setSort).not.toHaveBeenCalled();
    expect(fakeStore.mutations.setFilter).not.toHaveBeenCalled();
  });

  describe("sets up with url", () => {
    it.each`
      sortOrder    | sortProperty      | ratingFilter | genreFilter  | mutation       | payload
      ${true}      | ${"vote_average"} | ${undefined} | ${undefined} | ${"setSort"}   | ${{ property: "vote_average", order: true }}
      ${false}     | ${"vote_average"} | ${undefined} | ${undefined} | ${"setSort"}   | ${{ property: "vote_average", order: false }}
      ${true}      | ${"popularity"}   | ${undefined} | ${undefined} | ${"setSort"}   | ${{ property: "popularity", order: true }}
      ${undefined} | ${"vote_average"} | ${undefined} | ${undefined} | ${"setSort"}   | ${{ property: "vote_average", order: false }}
      ${true}      | ${undefined}      | ${undefined} | ${undefined} | ${"setSort"}   | ${{ property: "popularity", order: true }}
      ${undefined} | ${undefined}      | ${5}         | ${undefined} | ${"setFilter"} | ${{ type: "rating", value: 5 }}
      ${undefined} | ${undefined}      | ${undefined} | ${[0, 1]}    | ${"setFilter"} | ${{ type: "genre", value: 0 }}
      ${undefined} | ${undefined}      | ${undefined} | ${[0, 1]}    | ${"setFilter"} | ${{ type: "genre", value: 1 }}
    `(
      "set query ($sortOrder, $sortProperty, $ratingFilter, $genreFilter)",
      ({
        sortOrder,
        sortProperty,
        ratingFilter,
        genreFilter,
        mutation,
        payload
      }) => {
        jest.clearAllMocks();
        const query = { sortOrder, sortProperty, ratingFilter, genreFilter };
        wrapper = shallowMount(Home, { store, localVue, propsData: { query } });

        expect(fakeStore.mutations[mutation]).toHaveBeenCalledWith(
          expect.anything(),
          payload
        );
      }
    );
  });
});
