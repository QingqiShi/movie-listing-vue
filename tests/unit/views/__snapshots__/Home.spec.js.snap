// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue renders error 1`] = `
<div class="home">
  <action-bar-stub></action-bar-stub>
  <movie-list-stub></movie-list-stub>
  <error-modal-stub>test error</error-modal-stub>
</div>
`;

exports[`Home.vue renders no error 1`] = `
<div class="home">
  <action-bar-stub></action-bar-stub>
  <movie-list-stub></movie-list-stub>
  <!---->
</div>
`;
